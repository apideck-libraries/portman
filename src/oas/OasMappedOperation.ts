import { OpenAPIV3 } from 'openapi-types'

export interface IOasMappedOperation {
  id?: string
  path: string
  method: string
  pathRef: string
  schema: OpenAPIV3.OperationObject
  requestHeaders: OpenAPIV3.ParameterObject[]
  pathParams: OpenAPIV3.ParameterObject[]
  queryParams: OpenAPIV3.ParameterObject[]
  cookieParams: OpenAPIV3.ParameterObject[]

  requestBody(): OpenAPIV3.RequestBodyObject | undefined
}

export class OasMappedOperation implements IOasMappedOperation {
  public id?: string
  public path: string
  public method: string
  public pathRef: string
  public requestHeaders: OpenAPIV3.ParameterObject[]
  public pathParams: OpenAPIV3.ParameterObject[]
  public queryParams: OpenAPIV3.ParameterObject[]
  public cookieParams: OpenAPIV3.ParameterObject[]
  public schema: OpenAPIV3.OperationObject

  constructor(path: string, method: string, operation: OpenAPIV3.OperationObject) {
    this.schema = { ...operation }
    this.id = this.schema?.operationId
    this.method = method.toUpperCase()
    this.path = path
    this.pathRef = `${this.method}::${path}`
    this.requestHeaders = this.mapParameters('header')
    this.pathParams = this.mapParameters('path')
    this.queryParams = this.mapParameters('query')
    this.cookieParams = this.mapParameters('cookie')
  }

  private mapParameters(paramIn: string): OpenAPIV3.ParameterObject[] {
    if (!this.schema?.parameters) return []

    return (
      (this.schema.parameters as OpenAPIV3.ParameterObject[]).filter(
        parameter => (parameter.in as string) === paramIn
      ) || []
    )
  }

  requestBody(): OpenAPIV3.RequestBodyObject | undefined {
    return this.schema?.requestBody as OpenAPIV3.RequestBodyObject
  }

  requestBodySchema(mediaType: string): OpenAPIV3.SchemaObject | undefined {
    const requestBody = this.requestBody()
    if (!requestBody || !requestBody?.content?.[mediaType]?.schema) return

    return requestBody.content[mediaType].schema as OpenAPIV3.SchemaObject
  }
}
