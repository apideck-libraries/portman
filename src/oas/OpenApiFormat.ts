import { OpenAPIV3 } from 'openapi-types'
import openapiFormat, { OpenAPIFilterOptions, OpenAPIFilterSet } from 'openapi-format'

export interface IOpenApiFormatterConfig {
  inputFile: string
  filterFile: string
  outputFile?: string
}

export class OpenApiFormatter {
  public oas: OpenAPIV3.Document

  async filter(options: IOpenApiFormatterConfig): Promise<OpenAPIV3.Document> {
    const inputFilePath = options.inputFile
    const filterFilePath = options.filterFile
    const filterOptions = {} as OpenAPIFilterOptions

    // Load OpenAPI file
    this.oas = (await openapiFormat.parseFile(inputFilePath)) as unknown as OpenAPIV3.Document

    // Load filter file
    const filterSet = (await openapiFormat.parseFile(filterFilePath)) as OpenAPIFilterSet
    filterOptions.filterSet = filterSet

    // Filter OpenAPI file
    const resFilter = await openapiFormat.openapiFilter(this.oas, filterOptions)
    this.oas = resFilter.data as OpenAPIV3.Document

    // Write filtered OpenAPI file as YAML
    if (options.outputFile) {
      await openapiFormat.writeFile(options.outputFile, this.oas, { format: 'yaml' })
    }

    return this.oas
  }

  async parseFile(filePath: string): Promise<Record<string, unknown>> {
    return await openapiFormat.parseFile(filePath)
  }

  changeCase(valueAsString: string, caseType: string): string {
    return openapiFormat.changeCase(valueAsString, caseType)
  }
}
