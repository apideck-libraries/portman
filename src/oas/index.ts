export * from './OasMappedOperation'
export * from './OpenApiParser'
