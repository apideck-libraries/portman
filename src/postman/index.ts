export * from './PostmanMappedOperation'
export * from './PostmanParser'
