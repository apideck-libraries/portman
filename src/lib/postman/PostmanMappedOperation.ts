import { OasMappedOperation } from 'lib/oas/OasMappedOperation'
import { Item, Request } from 'postman-collection'

export interface IPostmanMappedOperation {
  id?: string
  path: string
  method: string
  pathRef: string
  pathRefVariable: string
  requestHeaders: Record<string, unknown>
  queryParams: Record<string, unknown>
  pathParams: Record<string, unknown>
  request: Request
  item: Item
}

export class PostmanMappedOperation implements IPostmanMappedOperation {
  public id?: string
  public path: string
  public method: string
  public pathRef: string
  public pathRefVariable: string
  public requestHeaders: Record<string, unknown>
  public queryParams: Record<string, unknown>
  public pathParams: Record<string, unknown>
  public request: Request
  public item: Item

  constructor(item: Item, request: Request, operationIdMap: Record<string, OasMappedOperation>) {
    this.item = item
    this.request = request
    this.method = request.method.toUpperCase()

    this.path = request.url.path ? `/${request.url.path.join('/')}` : '/'
    this.pathRef = this.normalizedPathRef(this.method)
    this.pathRefVariable = `${this.method}::${request.url.getPath()}`
    this.requestHeaders = request.headers.toJSON().map(({ key, value, description }) => {
      return { name: key, value, description: description?.content }
    })
    this.queryParams = request.url.query.map(({ key, value }) => {
      return { name: key, value }
    })
    this.pathParams = this.request.url.variables.toJSON().map(({ key, value, description }) => {
      return { name: key, value, description: description?.content }
    })

    this.id = operationIdMap[this.pathRef]?.id
  }

  private normalizedPathRef(method: string): string {
    const {
      request: { url }
    } = this
    const path = url?.path
      ?.map(segment => {
        return segment.includes(':') ? `{${segment}}`.replace(':', '') : segment
      })
      .join('/')
    return `${method}::/${path}`
  }
}
