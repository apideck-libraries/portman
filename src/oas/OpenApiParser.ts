import SwaggerParser from '@apidevtools/swagger-parser'
import { OpenAPIV3 } from 'openapi-types'
import path from 'path'
import { matchPath, METHODS } from 'utils'
import { OasMappedOperation } from './OasMappedOperation'

export interface OpenApiParserConfig {
  inputFile: string
}

export interface IOpenApiParser {
  oas: OpenAPIV3.Document
  mappedOperations: OasMappedOperation[]
  operationIdMap: Record<string, OasMappedOperation>
  getOperationById(operationId: string): OasMappedOperation | null
  getOperationByPath(path: string): OasMappedOperation | null
}

export class OpenApiParser {
  public oas: OpenAPIV3.Document
  public mappedOperations: OasMappedOperation[]
  public operationIdMap: Record<string, OasMappedOperation>

  async convert(options: OpenApiParserConfig): Promise<OpenAPIV3.Document> {
    const inputFile = path.resolve(options.inputFile)
    // Dereference the spec so all entities have encapsulated schemas
    const api = await SwaggerParser.dereference(inputFile)
    const oasDoc = (await SwaggerParser.bundle(api, {
      parse: {
        json: false, // Disable the JSON parser
        yaml: {
          allowEmpty: false // Don't allow empty YAML files
        }
      },
      dereference: {
        circular: false // Don't allow circular $refs
      },
      validate: {
        spec: false // Don't validate against the Swagger spec
      }
    })) as OpenAPIV3.Document

    this.oas = oasDoc
    this.pathsToOperations()
    this.mapOperationIds()
    return oasDoc
  }

  pathsToOperations = (): OasMappedOperation[] => {
    const paths = this.oas.paths
    const mappedOperations = Object.entries(paths)
      .filter(([_path, operations]) => !!operations)
      .map(([path, operations]) => {
        return (
          operations &&
          Object.entries(operations).map(([method, operation]): OasMappedOperation => {
            return new OasMappedOperation(path, method, operation as OpenAPIV3.OperationObject)
          })
        )
      })
      .reduce<OasMappedOperation[]>((acc, resource) => {
        resource && resource.map(item => acc.push(item))
        return acc
      }, [])

    this.mappedOperations = mappedOperations
    return this.mappedOperations
  }

  mapOperationIds = (): void => {
    const operationIdMap = this.mappedOperations.reduce((acc, operation) => {
      acc[operation.pathRef] = operation
      return acc
    }, {})

    this.operationIdMap = operationIdMap
  }

  public getOperationById(operationId: string): OasMappedOperation | null {
    return this.mappedOperations.find(({ id }) => id === operationId) || null
  }

  public getOperationByPath(path: string): OasMappedOperation | null {
    return this.mappedOperations.find(({ pathRef }) => pathRef === path) || null
  }

  public getOperationsByIds(operationIds: string[]): OasMappedOperation[] {
    return this.mappedOperations.filter(({ id }) => id && operationIds.includes(id))
  }

  public getOperationsByPath(path: string): OasMappedOperation[] {
    const targetSplit = path.split('::')
    const targetMethod = targetSplit[0].includes('*') ? METHODS : targetSplit[0]
    const targetPath = targetSplit[1]

    return this.mappedOperations.filter(mappedOperation => {
      return (
        targetMethod.includes(mappedOperation.method) && matchPath(targetPath, mappedOperation.path)
      )
    })
  }
}
