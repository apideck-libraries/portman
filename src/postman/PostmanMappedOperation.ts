import { Event, Item, ItemGroup } from 'postman-collection'
import { OasMappedOperation } from '../oas'

export type PostmanMappedOperationOptions = {
  item: Item
  operationIdMap?: Record<string, OasMappedOperation>
  id?: string
}

export class PostmanMappedOperation {
  public id?: string
  public path: string
  public method: string
  public pathRef: string
  public pathVar: string
  public requestHeaders: Record<string, unknown>
  public queryParams: Record<string, unknown>
  public pathParams: Record<string, unknown>
  public testJsonDataInjected: boolean
  public item: Item
  public postmanItemId?: string

  constructor(options: PostmanMappedOperationOptions) {
    const { item, operationIdMap, id } = options
    this.item = item
    const { request } = item

    this.method = request.method.toUpperCase()
    this.path = request.url.path ? `/${request.url.path.join('/')}` : '/'
    this.pathRef = this.normalizedPathRef(this.method)
    this.pathVar = this.normalizedPathVar(this.method)

    this.requestHeaders = request.headers.toJSON().map(({ key, value, description }) => {
      return { name: key, value, description: description?.content }
    })

    this.queryParams = request.url.query.map(({ key, value }) => {
      return { name: key, value }
    })

    this.pathParams = request.url.variables.toJSON().map(({ key, value, description }) => {
      return { name: key, value, description: description?.content }
    })

    this.id =
      !id && operationIdMap && operationIdMap[this.pathRef]?.id
        ? operationIdMap[this.pathRef]?.id
        : (id as string)

    this.postmanItemId = item?.id as string

    this.testJsonDataInjected = false
  }

  public getTests(): Event {
    return this.item.events.find(e => e?.listen === 'test', null)
  }

  getParent(): ItemGroup<Item> | null {
    const parent = this.item.parent()
    const isParent = ItemGroup.isItemGroup(parent)
    return isParent ? (parent as ItemGroup<Item>) : null
  }

  public getParentFolderId(): string | null {
    const parent = this.getParent()
    return parent ? parent.id : null
  }

  public getParentFolderName(): string | null {
    const parent = this.getParent()
    return parent ? parent.name : null
  }

  public clone({ newId, name }: { newId: string; name: string }): PostmanMappedOperation {
    // clone item for variation testing purposes
    // remove ids from item and response so we get a unique reference
    const clonedJsonItem = { ...this.item.toJSON() }
    const { id, ...clone } = clonedJsonItem

    if (name) {
      clone.name = name
      clone?.request ? (clone.request.name = name) : null
    }

    if (clone?.response?.length) {
      clone.response = clone.response.map(response => {
        const { id, ...clonedResponse } = response
        return clonedResponse
      })
    }
    const clonedPmItem = new Item(clone)
    clonedPmItem.events.clear()

    return new PostmanMappedOperation({
      item: clonedPmItem,
      id: newId
    })
  }

  private normalizedPathRef(method: string): string {
    const {
      item: {
        request: { url }
      }
    } = this

    const path = url?.path
      ?.map(segment => {
        return segment.includes(':') ? `{${segment}}`.replace(':', '') : segment
      })
      .join('/')
    return `${method}::/${path}`
  }

  private normalizedPathVar(method: string): string {
    const {
      item: {
        request: { url }
      }
    } = this

    const path = url?.path
      ?.map(segment => {
        return segment.includes(':') ? `{${segment}}`.replace(':', '') : segment
      })
      .join('/')
      .replace(/#|\//g, '-')
    return `${method.toLowerCase()}::${path}`
  }
}
