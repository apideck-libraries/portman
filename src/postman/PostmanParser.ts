import fs from 'fs-extra'
import path from 'path'
import { Collection, Item, ItemGroup, Request } from 'postman-collection'
import { matchPath, METHODS } from 'utils'
import { IOpenApiParser } from '../oas/OpenApiParser'
import { PostmanMappedOperation } from './PostmanMappedOperation'

export interface PostmanParserConfig {
  inputFile?: string
  postmanObj?: Collection
  oasParser?: IOpenApiParser
}

export interface IPostmanParser {
  collection: Collection
  mappedOperations: PostmanMappedOperation[]
  pmItems: Item[]
  getOperationById(operationId: string): PostmanMappedOperation | null
  getOperationByPath(path: string): PostmanMappedOperation | null
}

export class PostmanParser implements IPostmanParser {
  public collection: Collection
  public mappedOperations: PostmanMappedOperation[]
  public pmItems: Item[]

  private oasParser?: IOpenApiParser

  constructor(options: PostmanParserConfig) {
    const { inputFile, postmanObj, oasParser } = options
    this.oasParser = oasParser

    if (postmanObj) {
      this.collection = new Collection(postmanObj)
    } else if (inputFile) {
      const postmanJson = path.resolve(inputFile)
      this.collection = new Collection(JSON.parse(fs.readFileSync(postmanJson).toString()))
    }

    this.pmItems = []
    this.itemsToOperations()
  }

  getItems = (item: Item | ItemGroup<Item>): void => {
    if (Item.isItem(item)) {
      const request = (item as Item).request
      if (request && Request.isRequest(request)) {
        this.pmItems.push(item as Item)
      }
    }
    // Check if this is a nested folder
    else if (ItemGroup.isItemGroup(item)) {
      const items = (item as ItemGroup<Item>)?.items
      // Check if this is an empty folder
      if (items && items.count() === 0) return

      items.map(item => {
        return this.getItems(item)
      })
    }
  }

  itemsToOperations = (): void => {
    const { collection } = this
    collection?.items.map(item => {
      // Check if this is a request at the top level
      if (Item.isItem(item)) {
        return this.getItems(item)
      }
      // Check if this is a folder at the top level
      else if (ItemGroup.isItemGroup(item)) {
        const items = (item as ItemGroup<Item>).items

        items.each(item => {
          return this.getItems(item)
        })
      }
    })

    const operationIdMap = this.oasParser?.operationIdMap || {}
    this.mappedOperations = this.pmItems.map(item => {
      return new PostmanMappedOperation(item, operationIdMap)
    })
  }

  public getOperationByPath(path: string): PostmanMappedOperation | null {
    return this.mappedOperations.find(({ pathRef }) => pathRef === path) || null
  }

  public getOperationById(operationId: string): PostmanMappedOperation | null {
    return this.mappedOperations.find(({ id }) => id === operationId) || null
  }

  public getOperationsByIds(operationIds: string[]): PostmanMappedOperation[] {
    return this.mappedOperations.filter(({ id }) => id && operationIds.includes(id))
  }

  public getOperationsByPath(path: string): PostmanMappedOperation[] {
    const targetSplit = path.split('::')
    const targetMethod = targetSplit[0].includes('*') ? METHODS : targetSplit[0]
    const targetPath = targetSplit[1]

    return this.mappedOperations.filter(mappedOperation => {
      return (
        targetMethod.includes(mappedOperation.method) && matchPath(targetPath, mappedOperation.path)
      )
    })
  }
}
