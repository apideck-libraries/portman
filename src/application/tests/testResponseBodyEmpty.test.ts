import { getOasMappedOperation } from '../../../__tests__/testUtils/getOasMappedOperation'
import { getPostmanMappedOperation } from '../../../__tests__/testUtils/getPostmanMappedOperation'
import { OasMappedOperation } from '../../oas'
import { PostmanMappedOperation } from '../../postman'
import { testResponseBodyEmpty } from './testResponseBodyEmpty'

describe('testResponseBodyEmpty', () => {
  let oasOperation: OasMappedOperation
  let pmOperation: PostmanMappedOperation

  beforeEach(async () => {
    oasOperation = await getOasMappedOperation()
    pmOperation = await getPostmanMappedOperation()
  })

  it('should add test for empty body', async () => {
    pmOperation = testResponseBodyEmpty(pmOperation, oasOperation)
    const pmTest = pmOperation.getTests()
    expect(pmTest.script.exec).toMatchSnapshot()
  })
})
