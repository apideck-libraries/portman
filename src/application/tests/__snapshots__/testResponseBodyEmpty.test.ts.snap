// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testResponseBodyEmpty should add test for empty body 1`] = `
Array [
  "// Validate if response has empty Body 
pm.test(\\"[GET]::/crm/companies/:id - Response has empty Body\\", function () {
    pm.response.to.not.be.withBody;
});
",
]
`;
